### PR TITLE
ADEN-8905 Track adengine on/off.

### PR DESCRIPTION
**Pull request by jacekbj**
This ticket introduces a fundamental change into how AdEngine is loaded in Mobile-Wiki.

The main point is `routes/wiki-page.js`. This is the decision point for loading AdEngine.

If any part of Mobile-Wiki requires AdEngine loaded, it needs to use `Ads.getLoadedInstance()` method. It returns a Promise which:
* resolves when AdEngine is loaded
* rejects when a decision is made not to load AdEngine or AdEngine assets fail to load.

This way there is no risk of pending promises hanging on backend.

### CODE REVIEW

File: app/modules/ads/index.js [hunk 1]
```diff
@@ -154,11 +187,6 @@ class Ads {
   startAdEngine() {
     const { AdEngine } = window.Wikia.adEngine;
 
-    if (!this.isAdStackEnabled()) {
```
Comment (fraszczakszymon) [inline app/modules/ads/index.js:157]: Why is this removed? Smart banner relies on `fanTakeoverResolver` decision. Does it work properly now?

Comment (jacekbj) [reply app/modules/ads/index.js:157]: https://github.com/Wikia/mobile-wiki/blob/8b630f7e9a1f1678456a06ba402669cf07ef3ace/app/modules/ads/index.js#L336

`startAdEngine` is only run if `this.isAdStackEnabled()` is equal to true. Therefore this condition will never be fulfilled.

`fanTakeoverResolver` is resolved here: https://github.com/Wikia/mobile-wiki/blob/8b630f7e9a1f1678456a06ba402669cf07ef3ace/app/modules/ads/index.js#L184

Truth be told, I did not test this scenario when writing code.

Comment (fraszczakszymon) [reply app/modules/ads/index.js:157]: :facepalm:

Does it mean that smart banner is broken for some time already? It seems that fanTakeoverResolved isn't resolved at all when noads/ad stack is disabled :(

Comment (jacekbj) [reply app/modules/ads/index.js:157]: I fixed smart banner by resolving the promise if ad stack is disabled.

File: app/controllers/application.js [hunk 2]
```diff
@@ -44,10 +38,11 @@ export default Controller.extend(
...
 
     file: alias('lightbox.file'),
-    fullPage: oneWay('mobileApp'),
+    fullPage: oneWay('wikiPage.mobileApp'),
     isSearchPage: equal('currentRouteName', 'search'),
-    noAds: alias('ads.noAdsQueryParam'),
-    mobileApp: alias('ads.disableAdsInMobileApp'),
+    mobileApp: oneWay('wikiPage.mobileApp'),
+
+    ads: Ads.getInstance(),
```
Comment (fraszczakszymon) [inline app/controllers/application.js]: Seems not needed, is it?

File: app/modules/ads/index.js [hunk 3]
```diff
@@ -179,20 +142,14 @@ class Ads {
...
 
   createJWPlayerVideoAds(options) {
     const { jwplayerAdsFactory } = window.Wikia.adProducts;
 
-    if (this.isAdStackEnabled()) {
-      return jwplayerAdsFactory.create(options);
-    }
-
-    return null;
+    return jwplayerAdsFactory.create(options);
```
Comment (fraszczakszymon) [inline app/modules/ads/index.js:155]: What if there is no ad stack on FV page?

File: app/modules/ads/index.js [hunk 4]
```diff
@@ -46,16 +44,10 @@ class Ads {
       return adsPromise;
     }
 
-    if (typeof FastBoot !== 'undefined') {
```
Comment (fraszczakszymon) [inline app/modules/ads/index.js:49]: It's causing a memory leak on the server side, isn't it? Am I missing something?

File: app/components/ad-slot.js [hunk 5]
```diff
@@ -64,10 +57,6 @@ export default Component.extend(
       const ads = this.get('ads.module');
       const name = this.name;
 
-      if (this.noAds) {
-        return;
-      }
-
       if (this.insertOnViewportEnter) {
         ads.pushSlotToQueue(name);
```
Comment (fraszczakszymon) [inline app/components/ad-slot.js]: Hihi, this part doesn't make sense, right?
https://github.com/Wikia/mobile-wiki/pull/1821/files#diff-98564636413dff58d77852057b876c48R76-R77

File: app/modules/ads/index.js [hunk 6]
```diff
@@ -179,20 +142,14 @@ class Ads {
   finishFirstCall() {
     const { btfBlockerService } = window.Wikia.adEngine;
 
-    if (this.isAdStackEnabled()) {
-      btfBlockerService.finishFirstCall();
-      fanTakeoverResolver.resolve();
-    }
+    btfBlockerService.finishFirstCall();
```
Comment (fraszczakszymon) [inline app/modules/ads/index.js:145]: What if there is no ad stack on page?

Comment (fraszczakszymon) [reply app/modules/ads/index.js:145]: Shouldn't we mark it as private? (if it's executed in safe place)

File: app/modules/ads/index.js [hunk 7]
```diff
@@ -547,10 +489,6 @@ class Ads {
   waitForVideoBidders() {
     const { context, utils } = window.Wikia.adEngine;
 
-    if (!this.isAdStackEnabled()) {
```
Comment (fraszczakszymon) [inline app/modules/ads/index.js:564]: What if there is no ad stack? FV promise may not be resolved?

Comment (fraszczakszymon) [reply app/modules/ads/index.js:564]: Never mind, I've noticed the changes in `video-players/jwplayer.js`

File: app/modules/ads/index.js (hunk 3 above)
Comment (fraszczakszymon) [reply app/modules/ads/index.js:155]: Never mind, I've noticed your changes: 
https://github.com/Wikia/mobile-wiki/pull/1821/files#diff-1f8a4e7742e20ee4da173fdc8f902a2fR71

File: app/components/ad-slot.js [hunk 8]
```diff
@@ -40,10 +37,6 @@ export default Component.extend(
         return;
       }
 
-      if (this.noAds) {
```
Comment (fraszczakszymon) [inline app/components/ad-slot.js:43]: Do we need to push slot when ad stack is disabled?

File: app/modules/ads/index.js (hunk 1 above)
Comment (fraszczakszymon) [reply app/modules/ads/index.js:157]: Thank you

File: app/components/ad-slot.js (hunk 8 above)
Comment (jacekbj) [reply app/components/ad-slot.js:43]: Here I made sure that if ads are disabled then `Ads.waitForAdEngine` will not resolve
https://github.com/Wikia/mobile-wiki/pull/1821/commits/6c3ce1973e1d13a325e890b55e90b3e83608c1fc

Ads are injected after `Ads.waitForAdEngine` resolves, therefore there is no need to check for that in component:
https://github.com/Wikia/mobile-wiki/blob/ADEN-8905/app/components/article-content.js

File: app/components/ad-slot.js (hunk 5 above)
Comment (jacekbj) [reply app/components/ad-slot.js]: That's right, I removed `ads` service in favour of `Ads` module.
https://github.com/Wikia/mobile-wiki/pull/1821/commits/25acef63cb57ce43676d7621c6f965a638f071a4

File: app/controllers/application.js (hunk 2 above)
Comment (jacekbj) [reply app/controllers/application.js]: Removed.

File: app/modules/ads/index.js (hunk 4 above)
Comment (jacekbj) [reply app/modules/ads/index.js:49]: There should be no memory leaks now.

File: app/modules/ads/index.js (hunk 6 above)
Comment (jacekbj) [reply app/modules/ads/index.js:145]: Marked.

File: app/components/ad-slot.js (hunk 8 above)
Comment (fraszczakszymon) [reply app/components/ad-slot.js:43]: :+1:

File: app/components/ad-slot.js [hunk 9]
```diff
@@ -61,41 +52,41 @@ export default Component.extend(
...
-      } else if (this.shouldWaitForUapResponse) {
-        ads.waitForUapResponse(
-          () => ads.pushSlotToQueue(this.name),
-          () => {},
-        );
-      }
+      Ads.getLoadedInstance()
+        .then((ads) => {
+          if (this.insertOnViewportEnter) {
+            ads.pushSlotToQueue(this.name);
```
Comment (fraszczakszymon) [inline app/components/ad-slot.js:58]: `pushSlotToQueue` does not take any argument. Also, this if statement is broken (?). In this case, when `insertOnViewportEnter` is truthy, method `pushSlotToQueue` does not take any actions (returns void when `insertOnViewportEnter` is truthy). 🙈

Comment (fraszczakszymon) [reply app/components/ad-slot.js:58]: https://github.com/Wikia/mobile-wiki/pull/1821/files#diff-98564636413dff58d77852057b876c48R75

**Review by fraszczakszymon — APPROVED** [review]

File: app/components/ad-slot.js (hunk 9 above)
Comment (jacekbj) [reply app/components/ad-slot.js:58]: I tried to refactor this part but then got kind of dizzy looking at how convoluted this is. I will straighten this up.

Comment (jacekbj) [reply app/components/ad-slot.js:58]: My attempt to clean this place up deserves a ticket on its own.
Leaving it as is for now.
